Fix Windows MSVC AVX builds

cosmic-snow opened the pull request:
- A bug introduced in 0cb2b86730a56cee7460b0582418228e8049d3ef breaks AVX on Windows with MSVC
- currently getting: `warning C5102: ignoring invalid command-line macro definition '/arch:AVX2'`

## Describe your changes
- The solution is to use `_options(...)` not `_definitions(...)`

## Issue ticket number and link
- https://github.com/nomic-ai/gpt4all/issues/923

## Checklist before requesting a review
- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] I have added thorough documentation for my code.
- [ ] I have tagged PR with relevant project labels. I acknowledge that a PR without labels may be dismissed.
- [x] If this PR addresses a bug, I have provided both a screenshot/video of the original bug and the working solution.

## Demo
Description of the problem is in the accompanying issue: #923 

### Steps to Reproduce
Build on Windows without the fix produces warnings like:
```
command line : warning C5102: ignoring invalid command-line macro definition '/arch:AVX2' ...
```

cosmic-snow commented:
- Fixed: https://github.com/nomic-ai/gpt4all/issues/912#issuecomment-1585036035
- Probably also fixes: #936

rmcc3 reviewed (approved):
Here's a code review for the changes made in the pull request:

1. **Correctness**: The changes made in this pull request seem to be correct. The modifications made in the CMake file fix the issue with AVX builds on Windows with MSVC. The author has changed the `target_compile_definitions` to `target_compile_options` for MSVC builds using AVX512, AVX2, and AVX. The use of `target_compile_options` seems appropriate as we are dealing with compiler options and not preprocessor definitions. The changes made seem to correctly use the `/arch` flag which specifies the architecture for the MSVC compiler.

2. **Readability**: The changes do not impact the readability of the code. The code remains clear and easy to understand.

3. **Maintainability**: The changes do not introduce any additional complexity and should not impact the maintainability of the code. 

4. **Testing**: There is no evidence of new tests associated with these changes. Given the nature of the changes, it would be advisable to ensure that these changes are tested on the relevant platforms to ensure they work as intended.

5. **Documentation**: The changes are self-explanatory and do not require additional documentation. However, it might be helpful to include a comment indicating why the change from `target_compile_definitions` to `target_compile_options` was necessary for future reference. 

This pull request appears to be a valuable contribution to the project, improving its compatibility with Windows MSVC AVX builds. I recommend merging it after ensuring the changes have been adequately tested.

cosmic-snow commented:
@manyoso @niansa @AndriyMulyar 

I don't understand. The Windows MSVC build has had a bug for half a week. Why not accept the fix, or at least have a look?

I think I've followed all the requirements:
- opened a separate issue describing the problem and cause, plus where it got introduced
- added all the necessary information (or links) in this pull request

Later, I've also added two issues of other people affected by it, and then someone else even reviewed the fix.
Even before that, I tried to get your attention on the merge that broke it (partially so I wouldn't have to do the extra work).
<br>

Or can you give me some feedback, is something wrong with what I did/how I went about it? What am I missing here?
If it's because of the missing labels, I think so far I have not been able to add labels to anything.
<br>

P.S. To be entirely honest, I'm getting a big anxious about it. I really hope you're not doing another release before this is fixed, because it looks like the main build now uses MSVC.

rguo123 commented:
hey @cosmic-snow, just raised this PR with team. Thanks for contribution we are taking a look!

One note, looks like this PR was opened on Friday (EST) and it's the start of a new week. Developers are humans and have responsibilities outside of coding. Please have some more patience and mindfulness here.

manyoso commented:
cosmic sn

> @manyoso @niansa @AndriyMulyar
> 
> I don't understand. The Windows MSVC build has had a bug for half a week. Why not accept the fix, or at least have a look?
> 
> I think I've followed all the requirements:
> 
> * opened a separate issue describing the problem and cause, plus where it got introduced
> * added all the necessary information (or links) in this pull request
> 
> Later, I've also added two issues of other people affected by it, and then someone else even reviewed the fix. Even before that, I tried to get your attention on the merge that broke it (partially so I wouldn't have to do the extra work).
> 
> Or can you give me some feedback, is something wrong with what I did/how I went about it? What am I missing here? If it's because of the missing labels, I think so far I have not been able to add labels to anything.
> 
> P.S. To be entirely honest, I'm getting a big anxious about it. I really hope you're not doing another release before this is fixed, because it looks like the main build now uses MSVC.

Hi, i'm looking at it now and I think this is correct. Thank you! Thank you! Thank you!

My only question is why the patch doesn't fix all five cases?

cosmic-snow commented:
> hey @cosmic-snow, just raised this PR with team. Thanks for contribution we are taking a look!

One note, looks like this PR was opened on Friday (EST) and it's the start of a new week. Developers are humans and have responsibilities outside of coding. Please have some more patience and mindfulness here.

Yes I know, sorry about that. But I've seen commits between where the bug was introduced (and the PR) and now, so I was hoping someone would've noticed. And it really would kind of break things if you were to do a release.

> 
cosmic sn

@manyoso @niansa @AndriyMulyar

I don't understand. The Windows MSVC build has had a bug for half a week. Why not accept the fix, or at least have a look?

I think I've followed all the requirements:

opened a separate issue describing the problem and cause, plus where it got introduced
added all the necessary information (or links) in this pull request
Later, I've also added two issues of other people affected by it, and then someone else even reviewed the fix. Even before that, I tried to get your attention on the merge that broke it (partially so I wouldn't have to do the extra work).

Or can you give me some feedback, is something wrong with what I did/how I went about it? What am I missing here? If it's because of the missing labels, I think so far I have not been able to add labels to anything.

P.S. To be entirely honest, I'm getting a big anxious about it. I really hope you're not doing another release before this is fixed, because it looks like the main build now uses MSVC.

Hi, i'm looking at it now and I think this is correct. Thank you! Thank you! Thank you!

My only question is why the patch doesn't fix all five cases?

You mean these two? https://github.com/nomic-ai/gpt4all/blob/d03748cc15d57e040a246a249c917d14530ef9b3/gpt4all-backend/llama.cpp.cmake#L382-L390

I only fixed the changes introduced in https://github.com/nomic-ai/gpt4all/commit/0cb2b86730a56cee7460b0582418228e8049d3ef, and made the warnings I had go away (plus the other two `/arch:...`
I don't know whether the other two flags do what they're supposed to do, but those at least look like macro definitions.

manyoso commented:
Cool. I fixed the markdown issue btw with a different technique. Was that ok with you?

cosmic-snow commented:
> Cool. I fixed the markdown issue btw with a different technique. Was that ok with you?

Yes, I'm happy I don't have to mess with regexes in situations that are more appropriate for parsers. And playing around with Qt internals until they kind of do what's necessary is also not something I'd really want to spend time on.
